Cpm.py

```diff
@@ -4,43 +4,98 @@
+            if depending_activity in self.__project[activity]:
+                self._project[activity].append(depending_activity)
+                logger.debug("%s added to the activity", depending_activity)
+            else:
+                logger.debug("%s already in the list of depending activities", depending_activity)
         else:
-            self._project = {}
-        self._project_duration = 0
+            logger.debug("%s isn`t at the project", activity)
 
-    def add_activity(self, name, duration):
-        self._project[name] = Activity(name, duration)
-        logger.debug("%s added to the project", self._project[name])
+    def add_activity(self, activity):
+        # this method add activity to the project
+        if activity not in self.__project:
+            self._project[activity] = []
+            logger.debug("%s added to the project", activity)
+        else:
+            logger.debug("%s already at the", activity)
 
     def remove_activity(self, activity):
-        if activity in self._project:
-            self._project.pop(activity)
+        # this method removing activity from the project
+        if activity in self.__project:
+            self.__project.pop(activity)
             logger.debug("%s deleted from project", activity)
 
     def validate(self):
-        # todo create the validation
-        pass
+        # this method reveals a circle's activities in the project, and display them
+        for key, depending_activities in self.__project.items():
+            if key in depending_activities:
+                print("Activity : %s contains a circle", key)
+                logger.debug("Activity : %s contains a circle", key)
+                return False
+        return True  # No Circles!
 
     def find_isolated_activities(self):
-        # todo find isolated activities (An activity without following or ascending another activity)
+        #  find isolated activities (An activity without following or ascending another activity)
+        for key, depending_activities in self.__project.items():
+            if len(depending_activities) == 0:
+                print("Activity : %s is isolated activity", key)
+                logger.debug("Activity : %s is isolated activity", key)
+                return False
+        return True  # No Circles!
+
+    def calculate_activity_min_time_to_start(self):
+        for activity, depending_activity in self._project.items():
+            for dep_activity in depending_activity:
+                if (dep_activity.min_time_to_start == 0 or
+                        dep_activity.min_time_to_start > activity.duration + activity.min_time_to_start):
+                    dep_activity.min_time_to_start = activity.duration + activity.min_time_to_start
+                    dep_activity.max_time_to_start = dep_activity.min_time_to_start
+
+    def calculate_activity_max_time_to_start(self):
+        '''for activity, depending_activity in self._project.items():
```

Comment:
why you have the ''' at the start and at the end ?

Reply:
it is a comment for multiple line. its not good yet

Reply:
👍